polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -755,7 +755,7 @@ impl<Config: config::Config> XcmExecutor<Config> {
 						},
 					};
 				let actual_weight = maybe_actual_weight.unwrap_or(weight);
-				let surplus = weight.saturating_sub(actual_weight);
+				let surplus = require_weight_at_most.saturating_sub(actual_weight);
```

Comment:
I'm not entirely sure about this, and I'm just thinking out loud. From what I understand, it depends on the implementation details of [`FixedWeightBounds`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/weight.rs#L62-L66) and/or [`WeightInfoBounds`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/weight.rs#L108-L112), where `require_weight_at_most` is added to the instruction's weighed weight in the `fn instr_weight_with_limit`. It might also depend on the fact that `AllowTopLevelPaidExecutionFrom` utilizes the total calculated weights for `BuyExecution` [here](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/barriers.rs#L100).

In other words, if the runtime utilizes `FixedWeightBounds / WeightInfoBounds` and `AllowTopLevelPaidExecutionFrom`, then this fix should probably be ok and works. However, what about other runtimes that use different implementations without incorporating `require_weight_at_most` to the total instruction weight for `Transact`? Will this fix still be effective? Will it potentially refund less or more than expected?

I can see several tests for transacting here: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/tests/transacting.rs, but I think we should ensure coverage for the scenario described in the prdoc: "a very large `require_weight_at_most` then this would be non-refunded", both for the PAID and UNPAID methods in those tests. Currently, they all use `require_weight_at_most: Weight::from_parts(50, 50)`, which seems expected and pretty fine.

Reply:
Runtimes that do not utilize `require_weight_at_most` for the calculation of the total weight of the message should probably be more concerned about  potential DOS no? how else are you going to make sure the weight being bought is sufficient to cover for the execution?

Reply:
With respect to the tests, yes I can add them.

Reply:
I agree if you don't count `require_weight_at_most` then here you'd be refunding more than was paid for.
However, the most common way is charging for `require_weight_at_most` and I think most people's mental model is you get refunded this difference (`require_weight_at_most` - `actual_weight`) instead of the one that was being actually used.
It's also the situation described a little further in the comment.

If we don't want to do this we can leave it as is, but we should change the comment, communicate that you might not be refunded if you input a very large value in `require_weight_at_most` and make sure users try to get as close to the extrinsic `weight` as possible.